src/main/java/com/yahoo/memory/BaseWritableMemoryImpl.java

```diff
@@ -119,6 +127,80 @@ public void copyTo(final long srcOffsetBytes, final WritableMemory destination,
         dstOffsetBytes, lengthBytes);
   }
 
+  @Override
+  public void writeTo(final long offsetBytes, final long lengthBytes,
+      final WritableByteChannel out) throws IOException {
+    checkValidAndBounds(offsetBytes, lengthBytes);
+    if (unsafeObj instanceof byte[]) {
+      writeByteArrayTo((byte[]) unsafeObj, offsetBytes, lengthBytes, out);
+    } else if (unsafeObj == null) {
+      writeDirectMemoryTo(offsetBytes, lengthBytes, out);
+    } else {
+      // Memory is backed by some array that is not byte[], for example int[], long[], etc.
+      // We don't have other choice as to do extra intermediate copy.
+      writeToWithExtraCopy(offsetBytes, lengthBytes, out);
+    }
+  }
+
+  private void writeByteArrayTo(final byte[] unsafeObj, final long offsetBytes,
+      final long lengthBytes, final WritableByteChannel out) throws IOException {
+    final int off =
+        Ints.checkedCast(cumBaseOffset + offsetBytes - UnsafeUtil.ARRAY_BYTE_BASE_OFFSET);
+    final int len = Ints.checkedCast(lengthBytes);
+    final ByteBuffer bufToWrite = ByteBuffer.wrap(unsafeObj, off, len);
+    writeFully(bufToWrite, out);
+  }
+
+  private void writeDirectMemoryTo(final long offsetBytes, long lengthBytes,
+      final WritableByteChannel out) throws IOException {
+    long addr = getCumulativeOffset(offsetBytes);
+    // Do chunking, because it's likely that WritableByteChannel.write(ByteBuffer) in some network-
+    // or file-backed WritableByteChannel implementations with direct ByteBuffer argument could
+    // be subject of the same safepoint problems as in Unsafe.copyMemory and Unsafe.setMemory.
+    while (lengthBytes > 0) {
+      int chunk = (int) Math.min(CompareAndCopy.UNSAFE_COPY_MEMORY_THRESHOLD, lengthBytes);
+      ByteBuffer bufToWrite = wrap(addr, chunk);
+      writeFully(bufToWrite, out);
+      addr += chunk;
+      lengthBytes -= chunk;
+    }
+  }
+
+  /**
+   * This method is copied from https://github.com/odnoklassniki/one-nio/blob/
+   * 27c768cbd28ece949c299f2d437c9a0ebd874500/src/one/nio/mem/DirectMemory.java#L95
+   */
```

Comment:
Interesting.  This allows us to create a BB from an off-heap address and length. Cool!  Wondering if this might be useful to export.  I.e., given any chunk of off-heap < 2GB wrapped by Memory, we could export it as a BB.  I've been wary of having users ever pass around off-heap addresses, but this gets around that.

Reply:
I would not do that because the created ByteBuffer is safe to be used only as long as the MemoryHandle is not closed. I don't see how passing BB instead of raw address is better than passing the Memory itself.